File: JavaApp/src/main/java/com/charlezz/javaapp/feature/base/BaseAdapter.java
package com.charlezz.javaapp.feature.base;

import androidx.paging.PagedListAdapter;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;

public abstract class BaseAdapter<D, VH extends BaseViewHolder> extends PagedListAdapter<D, VH> {

    protected BaseAdapter(@NonNull DiffUtil.ItemCallback<D> diffCallback) {
        super(diffCallback);
    }

    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        int viewType = getItemViewType(position);
        int viewModelId = getViewModelId(viewType);
        holder.getBinding().setVariable(viewModelId, getItem(position));
        holder.getBinding().executePendingBindings();

    }

    @NonNull
    protected abstract int getViewModelId(int viewType);
}

File: KotlinApp/src/main/java/com/charlezz/finalarchitecture/feature/base/BaseAdapter.kt
package com.charlezz.finalarchitecture.feature.base

import android.arch.paging.PagedListAdapter
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.v7.util.DiffUtil
import android.view.LayoutInflater
import android.view.ViewGroup

abstract class BaseAdapter<T, VDB : ViewDataBinding, VH : BaseViewHolder<VDB>>
(val bindingResId:Int, diffCallback: DiffUtil.ItemCallback<T>)
    : PagedListAdapter<T, VH>(diffCallback) {
    abstract fun getViewHolderLayoutId():Int

    @Suppress("UNCHECKED_CAST")
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        val binding = DataBindingUtil.inflate<VDB>(
                LayoutInflater.from(parent.context),
                getViewHolderLayoutId(),
                parent,
                false)

        return object:BaseViewHolder<VDB>(binding) {} as VH
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        holder.binding.setVariable(bindingResId, getItem(position))
        holder.binding.executePendingBindings()
    }
}
